src/librustc_traits/lowering.rs

```diff
@@ -101,14 +101,15 @@ impl<'tcx> Lower<PolyDomainGoal<'tcx>> for ty::Predicate<'tcx> {
     }
 }
 
-/// Transforms an existing goal into a FromEnv goal.
-///
 /// Used for lowered where clauses (see rustc guide).
-trait IntoFromEnvGoal {
+trait IntoGoal {
```

Comment:
Nit: could we have two traits `IntoFromEnvGoal` and `IntoWellFormedGoal` instead of just one? I find the name `IntoGoal` not very informative.